File: src/main/java/com/flaghacker/sttt/common/Timer.java
package com.flaghacker.sttt.common;

public class Timer
{
	private final long time;

	private long start = -1;
	private boolean interrupted;

	public Timer(long time)
	{
		this.time = time;
	}

	public void start()
	{
		this.start = System.currentTimeMillis();
	}

	public boolean started()
	{
		return start != -1;
	}

	public void interrupt()
	{
		this.interrupted = true;
	}

	public boolean isInterrupted()
	{
		return interrupted;
	}

	public long timeLeft()
	{
		if (!started())
			throw new IllegalStateException("this Timer has not been started yet");

		long left = time - (System.currentTimeMillis() - start);
		return (left > 0) && !interrupted ? left : 0;
	}

	public boolean running()
	{
		return timeLeft() > 0;
	}
}

File: src/main/com/flaghacker/sttt/common/Timer.kt
package com.flaghacker.sttt.common

class Timer(private val length: Long) {
	private var start: Long? = null
	var isInterrupted: Boolean = false
	val running get() = timeLeft() > 0

	fun started() = start != null
	fun start() = apply {
		if (start != null)
			throw IllegalStateException("this Timer has already been started")

		start = System.currentTimeMillis()
	}

	fun interrupt() {
		this.isInterrupted = true
	}

	fun timeLeft(): Long {
		val start = start ?: throw IllegalStateException("this Timer hasn't been started yet")

		val left = length - (System.currentTimeMillis() - start)
		return if (left > 0 && !isInterrupted) left else 0
	}
}
